### ISSUE TITLE
Fortls 3.0 crashes

### ISSUE THREAD

**Issue by stefabat** [fortran-lang/fortls#398]
**Describe the bug**
Fortls 3.0 crashes while parsing a file in vscode, using the modern fortran extension. In version 2.13 it was unable to parse that file, but it was not crashing, such that I could still use the LSP in VS code.

**To Reproduce**
1. Get the file from https://github.com/cp2k/cp2k/blob/master/src/base/base_uses.f90
2. Run `fortls --debug_filepath base_uses.f90 --debug_parser`
3. Crashes because unable to parse line 42

The error is the following:
```
Traceback (most recent call last):
  File "/home/stefano/Miniconda/bin/fortls", line 8, in <module>
    sys.exit(main())
  File "/home/stefano/Miniconda/lib/python3.9/site-packages/fortls/__init__.py", line 21, in main
    debug_parser(args)
  File "/home/stefano/Miniconda/lib/python3.9/site-packages/fortls/debug.py", line 474, in debug_parser
    file_ast = file_obj.parse(debug=True, pp_defs=pp_defs, include_dirs=include_dirs)
  File "/home/stefano/Miniconda/lib/python3.9/site-packages/fortls/parsers/internal/parser.py", line 1331, in parse
    _, line, post_lines = self.get_code_line(
  File "/home/stefano/Miniconda/lib/python3.9/site-packages/fortls/parsers/internal/parser.py", line 1109, in get_code_line
    if FRegex.PP_ANY.match(next_line):
TypeError: expected string or bytes-like object
```
However, the error message i get from vscode is the following:
```
Error sending result: '<fortls.parsers.internal.parser.FortranFile object at 0x7a9448176d90>'. Reason: 'RecursionError('maximum recursion depth exceeded while pickling an object')'
```

**Expected behavior**
Either to parse the file, or not crashing if unable to parse it.

**Setup information (please complete the following information):**
 - OS: Arch Linux
 - Python Version 3.9 (from miniconda)
 - fortls Version 3.0
 - Code editor used: VS Code
 - Fortran extension: Modern Fortran v3.4.0 (pre-release) and v3.2.0

**Configuration information (please complete the following information):**
I use all the defaults in VS code - Modern Fortran.

**Comment by gnikit** [comment]
I don't think this worked with v2.13.0 either. Locally I it fails the same way you described. Are you using any options for `fortls`?

**Comment by gnikit** [comment]
This might be loosely related to this Issue https://github.com/fortran-lang/fortls/issues/236, where the preprocessor is supposed to `#include` Fortran AST nodes into a file, but it currently is unable to.

**Comment by stefabat** [comment]
Indeed the same problem occurs for v2.13.0, though, the vscode extension was able to connect to the language server and working properly despite the error (I just had a warning about the failed parsing of that particular file at every opening of the workspace).

I actually now reinstalled fortls v3.0.0 in a fresh conda environment and the behavior of the vscode extension (pointing to the freshly installed fortls) is the same as for v2.13.0, meaning that is again usable despite the failed parsing. Sorry for that, I don't know what went wrong with the update from v2.13.0 to v3.0.0 in the other conda environment.

In any case, continuing to debug this parsing issue: I don't use any special settings in the vscode extension, so I suppose that fortls is run without any additional option.
I tried to add the `--pp_suffixes .f90` option to get the file preprocessed, though, without luck.
If I use
```
fortls --pp_suffixes .f90 --debug_filepath base_uses.f90 --debug_rootpath . --debug_parser
```
the file is not preprocessed. If I change the extension to `.F90` it is preprocessed (without the `--pp_suffixes` option). So, I am not sure what is wrong there.

**Comment by gnikit** [comment]
I got to the bottom of this issue. I will issue a fix tomorrow and see if I can refactor the entire piece of code, since it is an absolute nightmare. The superficial reason is that the `&&` of the preprocessor is confused with the line continuation marker `&`. The underlying reason is linked to the Issue I posted before, where the preprocessor info need to better/safer be included in the syntax to be parsed by the fortls parser.
